### PR TITLE
Initial refactoring for phrase suggester

### PR DESCRIPTION
**Pull request by cbuescher**
Adding initial serialization methods (readFrom, writeTo) to the PhraseSuggestionBuilder, also adding the base test framework for serialiazation testing, equals and hashCode. Moving SuggestionBuilder out of the global SuggestBuilder for better readability.

Still missing: Handling of Smoothing Model (depends on https://github.com/elastic/elasticsearch/pull/16130) and DirectCandidateGenerator (depends on https://github.com/elastic/elasticsearch/pull/16185). Also Test needs to be extended to handle all parameters.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java [hunk 1]
```diff
@@ -255,8 +257,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
      * Sets the prefix to provide completions for.
      * The prefix gets analyzed by the suggest analyzer.
      */
+    @Override
     public CompletionSuggestionBuilder prefix(String prefix) {
-        super.setPrefix(prefix);
+        super.prefix(prefix);
```
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: Didn't we at some point want to move towards genuine getters and setters?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java [hunk 2]
```diff
@@ -362,4 +366,30 @@ protected XContentBuilder innerToXContent(XContentBuilder builder, Params params
...
     }
+
+    @Override
+    public String getWriteableName() {
+        return SUGGESTION_NAME;
+    }
+
+    @Override
+    public void doWriteTo(StreamOutput out) throws IOException {
+        // TODO
```
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: Shouldn't this have an implementation before going in - even if it's just some "not implemented exception" to catch coding errors until there's the real code?

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java [hunk 3]
```diff
@@ -221,4 +225,30 @@ public XContentBuilder innerToXContent(XContentBuilder builder, Params params) t
...
     }
+
+    @Override
+    public String getWriteableName() {
+        return SUGGESTION_NAME;
+    }
+
+    @Override
+    public void doWriteTo(StreamOutput out) throws IOException {
+        // TODO
```
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java]: See comments above.

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: Since almost all of the setters here are not get/set I wanted to make it consistent without too much changes. I'm happy to rename them all together in one go at a later stage though. Also if anybody else objects I can change this back of course, I just don't want to change all the non-get/set type ones at this point.

File: core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    protected SB serializedCopy(SB original) throws IOException {
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.writeSuggestion(original);;
+            try (StreamInput in = new NamedWriteableAwareStreamInput(StreamInput.wrap(output.bytes()), namedWriteableRegistry)) {
+                return (SB) in.readSuggestion();
+            }
+        }
+    }
+
+    protected static <T> void maybeSet(Consumer<T> consumer, T value) {
```
Comment (MaineC) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: I would have loved to have something like this for the *Sort tests and came up with a less elegant solution. Can we have that in some more generic helper class?

File: core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        if (randomBoolean()) {
+            consumer.accept(value);
+        }
+    }
+
+    /**
+     * helper to get a random value in a certain range that's different from the input
+     */
+    protected static <T> T randomValueOtherThan(T input, Supplier<T> randomSupplier) {
+        T randomValue = null;
```
Comment (MaineC) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: I would have loved to have something like this for the *Sort tests and came up with a less elegant solution. Can we have that in some more generic helper class?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java (hunk 2 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: @MaineC Exception is maybe a good idea. This however is PR against the feature branch, so this doesn't go into master and in order to share some framework code quickly we said we could merge this maybe with some open issues for now.

File: core/src/test/java/org/elasticsearch/search/suggest/CustomSuggesterSearchIT.java [hunk 6]
```diff
@@ -83,4 +76,52 @@ protected XContentBuilder innerToXContent(XContentBuilder builder, Params params
...
+            return builder;
+        }
+
+        @Override
+        public String getWriteableName() {
+            return "custom";
+        }
+
+        @Override
+        public void doWriteTo(StreamOutput out) throws IOException {
```
Comment (MaineC) [inline core/src/test/java/org/elasticsearch/search/suggest/CustomSuggesterSearchIT.java]: See comment about empty methods above

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java (hunk 1 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: Yeah - makes sense. Maybe track the final move to get/set in a separate issue?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java (hunk 2 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: My bad - didn't see this actually went against the feature branch. Sorry.

**Comment by MaineC** [comment]
Left a few comments. Feel free to ignore the ones about methods being empty. I didn't realize this wasn't going against master.

File: core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java (hunk 4 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: Sure, any idea for a good place and/or classname?

Comment (MaineC) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: Not one of my creative days today.

Maybe in the org.elasticsearch.test package, no idea what the classname could be :)

File: core/src/main/java/org/elasticsearch/search/suggest/phrase/PhraseSuggestionBuilder.java [hunk 7]
```diff
@@ -91,6 +113,13 @@ public PhraseSuggestionBuilder separator(String separator) {
     }
 
     /**
+     * get the maxErrirs setting
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/phrase/PhraseSuggestionBuilder.java]: should be "get token separator"?

File: core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+        }
+        if (regex != null) {
+            builder.field(REGEX_FIELD.getPreferredName(), regex);
+        }
+        builder.startObject(getSuggesterName());
+        if (analyzer != null) {
+            builder.field(ANALYZER_FIELD.getPreferredName(), analyzer);
+        }
+        if (fieldname != null) {
+            builder.field("field", fieldname);
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java]: Why not use `FIELDNAME_FIELD` here?

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java]: Forgot that, will do.

File: core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Base class for the different suggestion implementations.
+ */
+public abstract class SuggestionBuilder<T extends SuggestionBuilder<T>> extends ToXContentToBytes implements NamedWriteable<T> {
```
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java]: +1 to moving this to its own file

File: core/src/main/java/org/elasticsearch/search/suggest/phrase/PhraseSuggestionBuilder.java (hunk 7 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/phrase/PhraseSuggestionBuilder.java]: right, thanks

File: core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    protected abstract void doWriteTo(StreamOutput out) throws IOException;
+
+    @Override
+    public final boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
```
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java]: Shouldn't this check instead be:
`if (!(obj instanceof SuggestionBuilder))`

Because sub-classes of this class will call its equals to test the equality of its specific private fields.

File: core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+        T other = (T) obj;
+        return Objects.equals(name, other.name()) &&
+               Objects.equals(text, other.text()) &&
+               Objects.equals(prefix, other.prefix()) &&
+               Objects.equals(regex, other.regex()) &&
+               Objects.equals(fieldname, other.field()) &&
+               Objects.equals(analyzer, other.analyzer()) &&
+               Objects.equals(size, other.size()) &&
+               Objects.equals(shardSize, other.shardSize()) &&
+               doEquals(other);
```
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java]: I see, you are calling into the equals of the subclass.  My previous comment doesn't apply then.

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java (hunk 1 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: What was the motivation for wanting to make them genuine "get/set" methods?

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: @abeyad to my knowledge there was an ungoing debate. Last issue I can find is this https://github.com/elastic/elasticsearch/issues/14266 but it also states that it might be good to do this in one go.

File: core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java (hunk 4 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: Noted, but this will be a separate PR.

**Comment by cbuescher** [comment]
@areek @abeyad thanks for the review, I hope I adressed all of your comments, any objections about merging this then?

File: core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Base class for the different suggestion implementations.
+ */
+public abstract class SuggestionBuilder<T extends SuggestionBuilder<T>> extends ToXContentToBytes implements NamedWriteable<T> {
+
+    protected final String name;
+    // TODO this seems mandatory and should be constructor arg
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java]: ++

**Comment by abeyad** [comment]
@cbuescher LGTM ++

File: core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+            builder.field(SHARDSIZE_FIELD.getPreferredName(), shardSize);
+        }
+
+        builder = innerToXContent(builder, params);
+        builder.endObject();
+        builder.endObject();
+        return builder;
+    }
+
+    public String getSuggesterName() {
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java]: maybe we can make this private, I assume this is only for readability?

**Comment by areek** [comment]
@cbuescher LGTM. This looks great and thanks for the test infra :)

File: core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java (hunk 13 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/SuggestionBuilder.java]: Good point, will change that as well. And yes, currently only for readability. I hope we keep the same name constants as Id in the stream as well as for registering the parsers. That's what worked for queries at least.

**Comment by cbuescher** [comment]
Didn't merge this with master but with feature-suggest-refactoring branch in c00c0fa0.
